### PR TITLE
Fix commit status index problem

### LINKED ISSUE #14059: Commit build status not updating (intermittent)
- Gitea version (or commit ref): 1.14.0+dev-414-g4ef751215 
- Git version: 2.20.1
- Operating system: Debian based
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite

## Description

Occasionally one of my Drone builds will not update it's status properly, leaving the commit as "Build in progress" with the yellow circle, when actually checking Drone it shows "Successful", or "Failed".

This only happens occasionally, the only pattern I can see is its only been for PR's so far, and also when the server is under heavy load. I first noticed when I ran "update branch" on 4 PR's at the same time, each calls 2 builds, so I have many builds queued.

From the logs, the error seems to be a SQL duplicate value in the unique index.

```
[17/Dec/2020:09:27:30 +0000] "GET /api/v1/repos/xxxx/xxxx/raw/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8/.drone.yml HTTP/1.1" 200 2553 "-" "Go-http-client/1.1"
[17/Dec/2020:09:27:30 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 638 "-" "Go-http-client/1.1"
[17/Dec/2020:09:27:31 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 636 "-" "Go-http-client/1.1"
[17/Dec/2020:09:45:09 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 638 "-" "Go-http-client/1.1"
[17/Dec/2020:09:45:22 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 636 "-" "Go-http-client/1.1"
[17/Dec/2020:09:50:59 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 636 "-" "Go-http-client/1.1"
[17/Dec/2020:09:50:59 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 500 375 "-" "Go-http-client/1.1"
```

```
2020/12/17 09:50:59 ...pi/v1/repo/status.go:64:NewCommitStatus() [E] CreateCommitStatus: NewCommitStatus[repo_id: 1, user_id: 1, sha: a12437fc1f5b4153a3e25f9b8d82a39d
20b7e7e8]: Insert CommitStatus[/home/git/gitea-repositories/xxxx/xxxx.git, a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8]: Error 1062: Duplicate entry '5-1-a12437fc1f5b
4153a3e25f9b8d82a39d20b7e7e8' for key 'UQE_commit_status_repo_sha_index'
```

Looking at how the index is generated, I'm guessing two status updates came in simultaneously and both generated the same value? Maybe a lock on the db table is needed during index increment?

https://github.com/go-gitea/gitea/blob/4aabbacdfe2830e3b3808e42d6d89d6688754089/models/commit_status.go#L220-L240

~~I notice that file has been updated yesterday, but looking at the code it looks like the index is still generated the same way. I will update my gitea anyway and keep an eye on things.~~ Just happened again on latest master. I've updated Gitea version above.

### PR DESCRIPTION
**Pull request by lunny**
Replace #14160 and fix #14059

### CODE REVIEW

File: models/commit_status.go [hunk 1]
```diff
@@ -37,6 +37,94 @@ type CommitStatus struct {
 	UpdatedUnix timeutil.TimeStamp `xorm:"INDEX updated"`
 }
 
+// CommitStatusIndex represents a table for commit status index
+type CommitStatusIndex struct {
+	ID       int64
+	RepoID   int64  `xorm:"unique(repo_sha)"`
+	SHA      string `xorm:"unique(repo_sha)"`
+	MaxIndex int64  `xorm:"index"`
+}
```
Comment (zeripath) [inline models/commit_status.go]: Every table must have a primary key.

File: models/migrations/v195.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+func addTableCommitStatusIndex(x *xorm.Engine) error {
+	// CommitStatusIndex represents a table for commit status index
+	type CommitStatusIndex struct {
+		ID       int64
+		RepoID   int64  `xorm:"unique(repo_sha)"`
+		SHA      string `xorm:"unique(repo_sha)"`
+		MaxIndex int64  `xorm:"index"`
+	}
```
Comment (zeripath) [inline models/migrations/v195.go:20]: Every table must have a primary key

**Comment by zeripath** [comment]
Is the `ID` column is necessary? Could we use a composite pk?

File: models/commit_status.go (hunk 1 above)
Comment (lunny) [reply models/commit_status.go]: `ID int64` will become as primary key and autoincrement from `xorm`.

**Comment by lunny** [comment]
> Is the `ID` column is necessary? Could we use a composite pk?

Composite primary key makes things more complicated, I would like to add a `ID` column.

**Review by zeripath — APPROVED** [review]

**Review by zeripath — COMMENTED** [review]
ugh this is still failing tests.

**Comment by lunny** [comment]
@zeripath PASS now.

**Review by axifive — APPROVED** [review]

File: models/migrations/migrations.go [hunk 3]
```diff
@@ -342,6 +342,8 @@ var migrations = []Migration{
 	NewMigration("Add repo id column for attachment table", addRepoIDForAttachment),
 	// v194 -> v195
 	NewMigration("Add Branch Protection Unprotected Files Column", addBranchProtectionUnprotectedFilesColumn),
+	// v196 -> v197
```
Comment (delvh) [inline models/migrations/migrations.go:345]: Shouldn't that be
```suggestion
	// v195 -> v196
```
?

Comment (lunny) [reply models/migrations/migrations.go:345]: Ah, yes, typo

Comment (delvh) [reply models/migrations/migrations.go:345]: see #17130.

**Pull request merged**